### PR TITLE
[MIST-533] Create separate MISTMaster

### LINKED ISSUE #533: Make MIST master as a task
We want to change the MIST driver to deal with the query distribution regarding it's group. (#531)
Because of this, the bottleneck can be a bottleneck if a single driver does these things solely.
To distribute the MIST driver, we need to make MIST driver as a REEF task.

### PR DESCRIPTION
**Pull request by DifferentSC**
This PR resolves #533 and #423 via
* Creating a separate master
* Create avro protocols and setup connection among Driver, Master, and Tasks
* Implement tracking evaluator stats in MIST Master (#423)

Closes #533, #423.

### CODE REVIEW

File: mist-client/src/main/java/edu/snu/mist/client/MistClient.java [hunk 1]
```diff
@@ -30,7 +30,7 @@ private MistClient() {
   public static void main(final String[] args) throws IOException {
     if (args.length < 4) {
       System.out.println("Please enter the command-line parameters: ");
-      System.out.println("submit-jar $jar_path$ $driver_address$ $driver_port$");
+      System.out.println("submit-jar $jar_path$ $driver_address$ $driver_port");
```
Comment (taegeonum) [inline mist-client/src/main/java/edu/snu/mist/client/MistClient.java]: Why is it removed?

File: mist-common/src/main/avro/client_to_master_msg.avpr [hunk 2]
```diff
@@ -15,21 +15,25 @@
...
   "types":
   [
     {
       "name": "QueryInfo",
       "type": "record",
       "fields":
-      [ /* TODO[MIST-92]: Define avro RPC protocol between client and driver */
+      [
+        {
+          "name": "appName",
```
Comment (taegeonum) [inline mist-common/src/main/avro/client_to_master_msg.avpr]: appName -> appId

File: mist-common/src/main/avro/client_to_master_msg.avpr [hunk 3]
```diff
@@ -68,13 +72,9 @@
     "getTasks":
     {
       "request":
-      [
```
Comment (taegeonum) [inline mist-common/src/main/avro/client_to_master_msg.avpr]: Why is it removed?

File: mist-common/src/main/avro/task_to_master_msg.avpr [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /**
+  * This is a skeleton protocol for RPC communication from MistTask to MistMaster.
```
Comment (taegeonum) [inline mist-common/src/main/avro/task_to_master_msg.avpr]: add TODO?

File: mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import javax.inject.Inject;
+
+/**
+ * The tang configuration for master.
+ */
+public final class MistMasterConfigs {
+
+  /**
+   * Task information gathering term of MistMaster.
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java]: term -> period?

File: mist-core/src/main/java/edu/snu/mist/core/master/TaskLoadUpdater.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+   */
+  private final ProxyToTaskMap proxyToTaskMap;
+
+  /**
+   * The task-to-info map.
+   */
+  private final TaskInfoMap taskInfoMap;
+
+  public TaskLoadUpdater(final ProxyToTaskMap proxyToTaskMap,
+                          final TaskInfoMap taskInfoMap) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/master/TaskLoadUpdater.java]: Please align the line

File: mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java [hunk 7]
```diff
@@ -155,6 +154,10 @@ public QueryControlResult create(final Tuple<String, AvroDag> tuple) {
             new Object[]{appId, queryId});
       }
 
+      if (!applicationMap.containsKey(appId)) {
+        createApplication(appId, tuple.getValue().getJarPaths());
```
[the remark this answers is not in the document]
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java]: Addressed in 197d0db

File: mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  private final long taskInfoGatherPeriod;
+
+  @Inject
+  private MistMasterConfigs(
+      @Parameter(TaskInfoGatherTerm.class) final long taskInfoGatherPeriod
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java]: TaskInfoGatherTerm -> TaskInfoGatherPeriod?

File: mist-core/src/main/java/edu/snu/mist/core/rpc/DefaultClientToMasterMessageImpl.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+   */
+  private final TaskInfoMap taskInfoMap;
+
+  private final ApplicationCodeManager appCodeManager;
+
+  @Inject
+  private DefaultClientToMasterMessageImpl(
+      final TaskInfoMap taskInfoMap,
+      final ApplicationCodeManager appCodeManager
+  ) {
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/rpc/DefaultClientToMasterMessageImpl.java]: could you please remove the newline?

File: mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java [hunk 10]
```diff
@@ -155,8 +154,16 @@ public QueryControlResult create(final Tuple<String, AvroDag> tuple) {
...
+      }
+
       final ApplicationInfo applicationInfo = applicationMap.get(appId);
       final DAG<ConfigVertex, MISTEdge> configDag = configDagGenerator.generate(tuple.getValue());
+      // Waiting for group information being added
+      while (applicationInfo.getGroups().isEmpty()) {
+        // Do nothing
```
Comment (taegeonum) [inline mist-core/src/main/java/edu/snu/mist/core/task/groupaware/GroupAwareQueryManagerImpl.java]: How about adding `Thread.sleep(100)`?

File: mist-common/src/main/avro/task_to_master_msg.avpr (hunk 4 above)
Comment (DifferentSC) [reply mist-common/src/main/avro/task_to_master_msg.avpr]: Addressed.

File: mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java (hunk 5 above)
Comment (DifferentSC) [reply mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java]: Addressed

Comment (taegeonum) [reply mist-core/src/main/java/edu/snu/mist/core/driver/MistMasterConfigs.java]: gathering term -> gathering period?

**Comment by DifferentSC** [comment]
I have addressed your comments. Thanks a lot, @taegeonum!

**Comment by taegeonum** [comment]
LGTM. I'm merging it

**Pull request merged**